.github/workflows/test-and-deploy.yml

```diff
@@ -0,0 +1,163 @@
+    name: Test & Deploy
+    needs: python_version
+    runs-on: ubuntu-latest
+    concurrency: test-deploy-${{ github.repository }}
+
+    env:
+      SLS_DEBUG: "*"
+
+    defaults:
+      run:
+        working-directory: ${{ inputs.working_directory }}
+
+    steps:
+      - uses: actions/checkout@v3
+
+      - uses: webfactory/ssh-agent@v0.5.4
+        with:
+          ssh-private-key: ${{ secrets.SSH_PRIVATE_KEY }}
+
+      - name: Install poetry
+        run: pipx install poetry
+
+      - uses: actions/setup-python@v4
+        with:
+          python-version: ${{ needs.python_version.outputs.version }}
+          cache: 'poetry'
+
+      # Poetry must explicitly be told to use the expected version https://github.com/actions/setup-python/issues/374
+      - name: Install dependencies
+        run: |
+          poetry env use ${{ needs.python_version.outputs.version }}
+          poetry config http-basic.properly "${{ secrets.PIP_GEMFURY_DEPLOY_TOKEN }}" NOPASS
+          poetry install
+          npm install
+
+      - name: Lint
+        run: poetry run inv lint
+
+      - name: Typecheck
+        if: inputs.run_typecheck 
+        run: poetry run inv typecheck
+
+      - name: Run Tests
+        run: poetry run inv test --junitxml="tests/results.xml"
+
+      - name: Publish test results
+        uses: EnricoMi/publish-unit-test-result-action@v1
+        if: always()
+        with:
+          files: "${{ inputs.working_directory }}/tests/*.xml"
+          comment_mode: "always"
+
+      - name: Upload metadata to Datadog
+        if: inputs.run_datadog_ci
+        run: poetry run inv datadog-ci
+        env:
+          DATADOG_API_KEY: ${{ secrets.DD_API_KEY }}
+
+      - name: Start deployment (staging)
+        id: deployment_staging
+        if: ${{ github.ref == format('refs/heads/{0}', inputs.default_branch) && inputs.deploy_staging }}
+        uses: bobheadxi/deployments@v1
+        with:
+          step: start
+          token: ${{ secrets.GITHUB_TOKEN }}
+          env: staging
+
+      - name: Create domain (staging)
+        if: ${{ github.ref == format('refs/heads/{0}', inputs.default_branch) && inputs.deploy_staging }}
+        run: npx sls create_domain --stage staging
+        env:
+          AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          DD_API_KEY: ${{ secrets.DD_API_KEY }}
+
+      - name: Deploy (staging)
+        if: ${{ github.ref == format('refs/heads/{0}', inputs.default_branch) && inputs.deploy_staging }}
+        run: poetry run inv deploy --stage staging
+        env:
+          AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          DD_API_KEY: ${{ secrets.DD_API_KEY }}
+
+      - name: Update deployment status (staging)
+        if: ${{ always() && github.ref == format('refs/heads/{0}', inputs.default_branch) && inputs.deploy_staging }}
```

Comment:
what is `always()` doing here? I thought it was for conditionally running the step. If it always returns true, why include it at all?

Reply:
Good question.. `always()` is used to run a step even if any previous steps fail. So here, even if the deployment fails, we want to update the deployment status to inform the user whether it passed or failed.
This is the same reason why we have included it in `publish test results` step, so that even if some of the tests fail, we still publish the results for the user.